include/std/experimental/conqueue

```diff
@@ -437,6 +462,12 @@ template <typename T> struct buffer_queue<T>::pop_sender {
             "async_pop: queue is empty, putting %p in the waiters queue\n",
             &op);
         self.waiters.push_back(&op);
+        if constexpr (sp) {
+          // Now we are in the queue, set up cancellation callback.
+          if (stop_possible) {
+            op.stop_callback.emplace(stop_token, cancel_callback{op});
```

Comment:
Registering the calllback might end up executing the cancel_callback inline here while you still hold the lock which will try to recursively lock the callback on the same thread.

You need to either release the lock before registering the callback, or use an additional atomic to notify the callback when you've finished constructing it and it is now free to deregister the callback.

This is actually a bit of a flaw in the design of stop_callback. Ideally it would have passed a boolean to the callback that indicated whether it was being called inline in the constructor or asynchronously inside a call to request_stop().

Reply:
Thank you! Addressed in: https://github.com/GorNishanov/conqueue/pull/3